### PR TITLE
docs: Flesh out openapi for tag: Environments

### PR DESCRIPTION
**Pull request by chriswk**
### What
This fleshes out the documentation for actions tagged with Environments. 

A rather small PR since previous efforts had covered most of the operations already.

### CODE REVIEW

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 1]
```diff
@@ -8,22 +8,37 @@ export const environmentProjectSchema = {
...
         name: {
             type: 'string',
+            example: 'development',
+            description: 'The name of the environment',
         },
         type: {
             type: 'string',
+            example: 'production',
+            description:
+                'The type of environment. Typically either development (non-production) or production',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:18]: A couple things:
1. We should link to environment types here (added in the suggestion).
2. Is this a "read" schema / is it returned by a GET call? If so, I don't think we need the sentence about what it probably is. (removed in suggestion)

```suggestion
                'The [type of environment](https://docs.getunleash.io/reference/environments#environment-types).',
```

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 2]
```diff
@@ -8,22 +8,37 @@ export const environmentProjectSchema = {
...
             type: 'boolean',
+            example: true,
+            description:
+                '`true` if the environment is enabled for the project, otherwise `false`',
         },
         protected: {
             type: 'boolean',
+            example: false,
+            description:
+                '`true` if the environment is not removable, otherwise `false`',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:30]: What does removable mean here? Does that mean that it can't be deleted? If so, how about something like this?

```suggestion
                '`true` if the environment is protected, otherwise `false`. A *protected* environment can not be deleted.',
```

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 3]
```diff
@@ -8,22 +8,37 @@ export const environmentProjectSchema = {
...
         },
         sortOrder: {
             type: 'number',
+            example: 1,
+            description:
+                'Priority of the environment in a list of environments, the lower value, the higher up in the list the environment will appear',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:36]: I think I've asked this before, but there's no minimum or maximum for this right? Negative numbers are also okay? Do they have to be integers?

```suggestion
            example: 1,
            description:
                'Priority of the environment in a list of environments, the lower the value, the higher up in the list the environment will appear',
```

File: src/lib/openapi/spec/environment-project-schema.ts
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:1]: I'd like the schema itself to have a description as well, and not just its fields. What does the schema describe?

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 4]
```diff
@@ -8,22 +8,37 @@ export const environmentProjectSchema = {
...
         },
         projectApiTokenCount: {
             type: 'number',
             nullable: true,
+            description: 'How many API tokens are register',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:41]: So this field can be a number OR null OR undefined (equivalent to `number | null | undefined` in TS)? When is it anything but a number? That should be mentioned in the description.

Also, I'm not sure I understand the description here. What does it mean? And what kind of API tokens does it count?

This suggestion needs to be reworked before it's ready.

```suggestion
            nullable: true,
            description: 'The number of client API tokens that have access to this project. `null` if .... `undefined` if ...'.
```

File: src/lib/openapi/spec/environment-schema.ts [hunk 5]
```diff
@@ -25,12 +25,14 @@ export const environmentSchema = {
         },
         protected: {
             type: 'boolean',
+            example: true,
+            description: '`true` if the environment is a protected environment',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-schema.ts:29]: As in the previous comment. What is a protected environment? If this shows up in a lot of places, we should add it to the docs. I couldn't find anything about protected environments in the docs, so we should probably add it anyway.

File: src/lib/routes/admin-api/environments.ts [hunk 6]
```diff
@@ -73,8 +80,10 @@ export class EnvironmentsController extends Controller {
                 openApiService.validPath({
                     tags: ['Environments'],
                     operationId: 'getEnvironment',
+                    description: 'Gets the environment with `name`',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:83]: This is the environment ID, right? Or is it its name? We should also have a summary.

```suggestion
                    summary: 'Get environment by ID',
                    description: 'Gets the environment whose ID is `name`',
```

File: src/lib/routes/admin-api/environments.ts [hunk 7]
```diff
@@ -58,8 +61,12 @@ export class EnvironmentsController extends Controller {
             middleware: [
                 openApiService.validPath({
                     tags: ['Environments'],
+                    description: 'Gets all environments',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:64]: This should probably be a summary and not a description.

We haven't discussed this previously and it's not written down (I'll do it 🙋🏼), but summaries should probably be expressed as a "bare infinitive", as in "call this endpoint to [summary]".



```suggestion
                    summary: 'Get all environments',
                    description: 'Retrieves all environments that exist in this Unleash instance.'
```

File: src/lib/routes/admin-api/environments.ts [hunk 8]
```diff
@@ -89,8 +98,11 @@ export class EnvironmentsController extends Controller {
                 openApiService.validPath({
                     tags: ['Environments'],
                     operationId: 'getProjectEnvironments',
+                    description:
+                        'Gets the environments that are available for this project',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:102]: Something like this, maybe? What makes an environment available?

```suggestion
                    summary: 'Get environments available to a project',
                    description:
                        'Gets all environments that are available for the specified project. An environment is considered available if ... what?',
```

File: src/lib/routes/admin-api/environments.ts [hunk 9]
```diff
@@ -104,9 +116,14 @@ export class EnvironmentsController extends Controller {
             middleware: [
                 openApiService.validPath({
                     tags: ['Environments'],
+                    description:
+                        'Updates sort orders for the named environments',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:120]:
```suggestion
                    summary: 'Update environment sort orders'
                    description:
                        'Updates sort orders for the named environments. Environments not specified are unaffected.',
```

File: src/lib/routes/admin-api/environments.ts [hunk 10]
```diff
@@ -120,8 +137,12 @@ export class EnvironmentsController extends Controller {
             middleware: [
                 openApiService.validPath({
                     tags: ['Environments'],
+                    description: 'Toggles the environment with `name` on',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:140]: Is this enable/disable environments for the global unleash instance or for projects? If the former, is it whether an env should be deprecated or not? 

I think the description could be a bit clearer on what this does. The operation is also missing a summary.

File: src/lib/routes/admin-api/environments.ts [hunk 11]
```diff
@@ -135,8 +156,12 @@ export class EnvironmentsController extends Controller {
             middleware: [
                 openApiService.validPath({
                     tags: ['Environments'],
+                    description: 'Toggles the environment with `name` off',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/environments.ts:159]: Same comment here as for the `on` endpoint above: there's no summary and I'm don't understand what turning an environment on or off means.

**Review by thomasheartman — CHANGES_REQUESTED** [review]
Ayy, that was fast 🥳 Nice work, but I have a few suggestions, a few questions, and a few recommendations 😄 If you're wondering about anything, please let me know. I'll go ahead and update the parent task in the meantime

File: src/lib/openapi/spec/environment-project-schema.ts (hunk 2 above)
Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:30]: Agreed.

File: src/lib/openapi/spec/environment-project-schema.ts (hunk 3 above)
Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:36]: No min or max, should be integers yes, I don't think we accept decimals.

File: src/lib/openapi/spec/environment-project-schema.ts
Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:1]: Agreed. Will fix

File: src/lib/openapi/spec/environment-project-schema.ts (hunk 3 above)
Comment (thomasheartman) [reply src/lib/openapi/spec/environment-project-schema.ts:38]: Nice, thanks! In that case, I think there's only a missing "the" in the sentence (as mentioned in the suggestion) 😄

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 12]
```diff
@@ -5,29 +5,51 @@ export const environmentProjectSchema = {
...
+                'Priority of the environment in a list of environments, the lower value, the higher up in the list the environment will appear',
         },
         projectApiTokenCount: {
-            type: 'number',
-            nullable: true,
+            type: 'integer',
+            minimum: 0,
+            example: 5,
+            description:
+                'How many client and frontend API tokens that have access to this project',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:45]: Great! I suggest rewording it to the suggestion below, but I'll leave it up to you.


```suggestion
                'The number of client and front-end API tokens that have access to this project',
```

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 13]
```diff
@@ -5,29 +5,51 @@ export const environmentProjectSchema = {
     type: 'object',
     additionalProperties: false,
     required: ['name', 'type', 'enabled'],
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:7]: So here's a thing to discuss. The way OpenAPI works, **all** properties that aren't specified in the `required` list are considered optional. From [the OpenAPI docs](https://swagger.io/docs/specification/data-models/data-types/#required):

>  By default, all object properties are optional. You can specify the required properties in the required list: 

That means, they can be undefined (as in: not included in the object). My guess is that most of the other fields on this schema aren't optional: they're always there. Can we specify all the properties that aren't optional in this list. And if a property really is optional, then its description should explicitly state when it will be missing. What do you think?

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 14]
```diff
@@ -5,29 +5,51 @@ export const environmentProjectSchema = {
...
+    description:
+        "Describes an environment as it's configured for the project it's connected to",
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:9]: Non-blocking and up to you, but I feel like this is actually describing a project and not an environment. Maybe we can flip it on its head? It's a tricky thing to explain, this 😅 

```suggestion
        "Describes a project's configuration in a given environment.",
```

File: src/lib/openapi/spec/environment-project-schema.ts [hunk 15]
```diff
@@ -5,29 +5,51 @@ export const environmentProjectSchema = {
...
+            description:
+                'How many client and frontend API tokens that have access to this project',
         },
         projectEnabledToggleCount: {
             type: 'number',
-            nullable: true,
+            minimum: 0,
+            example: 7,
+            description:
+                'Number of features enabled in this environment for this project',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environment-project-schema.ts:52]: Missing a definite article here

```suggestion
                'The number of features enabled in this environment for this project',
```

File: src/lib/openapi/spec/environments-project-schema.ts [hunk 16]
```diff
@@ -6,6 +6,7 @@ export const environmentsProjectSchema = {
     type: 'object',
     additionalProperties: false,
     required: ['version', 'environments'],
+    description: 'Environments defined for the current project',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/environments-project-schema.ts:9]: I'd switch this to something like this:

```suggestion
    description: 'Environments defined for a given project',
```

There's not necessarily a "current" project when you're working with APIs or reading specs, so I think this is clearer.

**Review by thomasheartman — COMMENTED** [review]
Good going! 🙌🏼  Looks like you're still looking through some of the comments I left, but I had another look now. I think an important thing that we should get out of the way is how `required` properties work in OpenAPI (refer to the comment about that specifically): everything not mentioned explicitly can be undefined according to the spec. 

Let's figure out how we want to treat that.

File: src/lib/openapi/spec/environment-project-schema.ts (hunk 13 above)
Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:7]: I think you're right, since this is a response schema, not a request schema, it should be safe to mark them all as required. I'll update the PR

Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:7]: Hmm, interesting discussion. The types we're returning here has these fields set as optional, so should we guarantee that they are here using required and change the Typescript type as well?

Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:7]: I see we have a call for Environments which does not return counts.

Comment (chriswk) [reply src/lib/openapi/spec/environment-project-schema.ts:7]: I'm guessing this is an instance of reusing types multiple places, where what we'd like is at least two different types, one with counts and one without, instead of trying to reuse the same type by making the fields optional

**Review by thomasheartman — APPROVED** [review]
Perfect! Yeah, this looks great to me for now 😄 I agree with the separate PR to split out the environment schemas from the other schemas, so let's get this done for now 💯
